### PR TITLE
Update dataset api and docs

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
Currently the Dataset python API has four initializers exposed. These are:

1. Dataset(arg0: Dict[str, DataArray])
2. Dataset(arg0: DataArray)
3. Dataset(data: Dict[str, Variable]={}, coords: Dict[str, Variable]={})
4. Dataset(arg0: Dataset)

This PR removes 2 and 4 from this list to simplify the API. The reasoning for this is.
* For 4 this is used to convert a view into a full Dataset or copy an existing dataset. This can already be achived by the .copy() methods on dataset.
* For 2 this can be easily achieved using 1 and is I think a rare case which obfuscates the fact that Dataset is a Dict of DataArrays slightly. This is a more marginal removal I have done so for now so people can see how it works in practice.

The Docstrings have for the init method have also been updated. I decided to manually specify the available overloads in the docstring as the auto-generated ones were unclear and hard to read. I've added a screenshot of the new docs page below. This still has the overloaded function message but whilst we have two incompatible overloads this feels unavoidable.

image

### CODE REVIEW

File: python/dataset.cpp [hunk 1]
```diff
@@ -281,20 +281,41 @@ void init_dataset(py::module &m) {
...
+             Each Variable is converted into a DataArray within the Dataset.
+             )")
+      .def(py::init<const std::map<std::string, DataArrayConstView> &>(),
+           py::arg("data"),
+           R"(Overloaded Function
+           
+              __init__(self, data: Dict[str, DataArrayConstView])
+              
+              Initialise from a dictionary of DataArrays.
+              )");
```
Comment (SimonHeybrock) [inline python/dataset.cpp:315]: http://xarray.pydata.org/en/stable/generated/xarray.Dataset.html accepts both variables and data arrays in the same constructor. Can we do that?

Comment (Matthew-Andrew) [reply python/dataset.cpp:315]: Yes probably can I'll have a look into it. Two questions:
* Do we want to allow a dict made up of a mix of DataArrays and Variables or restrict them to only one.
* How do we want to handle the case where the user passed a dict of DataArrays and a dict of coords? This is currently not an allowed overload but could it be?

Comment (SimonHeybrock) [reply python/dataset.cpp:315]: In pricple the mix is ok, I think. Maybe this constructor can be written on the Python side and simply use something like:
```python
ds = sc.Dataset()
for coord in coords:
 ds.coord[coord] = ...
for data in data:
 ds[data] = ...
```
Then your two questions would be handled automatically?

**Comment by Matthew-Andrew** [comment]
I've updated the dataset python overload so that it can take a map containing either a Variable or a DataArray. It then constructs the relevant dataset from this. 

The documentation page now looks like.

image

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**